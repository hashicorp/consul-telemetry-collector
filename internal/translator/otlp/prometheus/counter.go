package prometheus

import (
	prompb "github.com/prometheus/client_model/go"
	"go.opentelemetry.io/collector/pdata/pmetric"
)

// AddCounter converts a prometheus counter to an OTLP monotonic Sum and adds it to the metrics builder
func (b *Builder) AddCounter(family *prompb.MetricFamily) {
	otlpMetric := pmetric.NewMetric()

	otlpMetric.SetName(normalizeName(family.GetName()))
	otlpMetric.SetDescription(family.GetHelp())
	emptySum := otlpMetric.SetEmptySum()
	emptySum.SetIsMonotonic(true)
	emptySum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
	for _, metric := range family.GetMetric() {
		dp := emptySum.DataPoints().AppendEmpty()

		for _, labelPair := range metric.GetLabel() {
			dp.Attributes().PutStr(labelPair.GetName(), labelPair.GetValue())
		}

		dp.SetTimestamp(timestampFromMs(metric.GetTimestampMs()))
		dp.SetDoubleValue(metric.GetCounter().GetValue())
	}

	b.metrics = append(b.metrics, otlpMetric)
}
