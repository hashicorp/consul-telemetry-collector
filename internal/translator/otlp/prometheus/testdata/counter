[
  {
    "name": "cluster.upstream_cx_destroy",
    "type": 0,
    "metric": [
      {
        "label": [
          {
            "name": "envoy.cluster_name",
            "value": "metrics_cluster"
          }
        ],
        "counter": {
          "value": 4
        },
        "timestamp_ms": 1682099630190
      }
    ]
  },
  {
    "name": "cluster.upstream_cx_destroy",
    "type": 0,
    "metric": [
      {
        "label": [
          {
            "name": "envoy.cluster_name",
            "value": "service_envoyproxy_io"
          }
        ],
        "counter": {
          "value": 23
        },
        "timestamp_ms": 1682099630190
      }
    ]
  },
  {
    "name": "http.downstream_rq_total",
    "type": 0,
    "metric": [
      {
        "label": [
          {
            "name": "envoy.http_conn_manager_prefix",
            "value": "admin"
          }
        ],
        "counter": {
          "value": 3
        },
        "timestamp_ms": 1682099630190
      }
    ]
  },
  {
    "name": "http.downstream_rq_total",
    "type": 0,
    "metric": [
      {
        "label": [
          {
            "name": "envoy.http_conn_manager_prefix",
            "value": "ingress_http"
          }
        ],
        "counter": {
          "value": 43
        },
        "timestamp_ms": 1682099630190
      }
    ]
  }
]
