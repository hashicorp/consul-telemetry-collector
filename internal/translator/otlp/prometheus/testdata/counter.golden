{
  "resourceMetrics": [
    {
      "resource": {},
      "scopeMetrics": [
        {
          "scope": {},
          "metrics": [
            {
              "name": "http_downstream_rq_total",
              "sum": {
                "dataPoints": [
                  {
                    "attributes": [
                      {
                        "key": "envoy_http_conn_manager_prefix",
                        "value": {
                          "stringValue": "admin"
                        }
                      }
                    ],
                    "startTimeUnixNano": "1682099630856000000",
                    "timeUnixNano": "1682099630856000000",
                    "asDouble": 4
                  },
                  {
                    "attributes": [
                      {
                        "key": "envoy_http_conn_manager_prefix",
                        "value": {
                          "stringValue": "ingress_http"
                        }
                      }
                    ],
                    "startTimeUnixNano": "1682099630856000000",
                    "timeUnixNano": "1682099630856000000",
                    "asDouble": 43
                  }
                ],
                "aggregationTemporality": 2,
                "isMonotonic": true
              }
            },
            {
              "name": "cluster_upstream_cx_destroy",
              "sum": {
                "dataPoints": [
                  {
                    "attributes": [
                      {
                        "key": "envoy_cluster_name",
                        "value": {
                          "stringValue": "metrics_cluster"
                        }
                      }
                    ],
                    "startTimeUnixNano": "1682099630856000000",
                    "timeUnixNano": "1682099630856000000",
                    "asDouble": 4
                  },
                  {
                    "attributes": [
                      {
                        "key": "envoy_cluster_name",
                        "value": {
                          "stringValue": "service_envoyproxy_io"
                        }
                      }
                    ],
                    "startTimeUnixNano": "1682099630856000000",
                    "timeUnixNano": "1682099630856000000",
                    "asDouble": 23
                  }
                ],
                "aggregationTemporality": 2,
                "isMonotonic": true
              }
            }
          ]
        }
      ]
    }
  ]
}
